04-onion-routing.md

```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Route Blinding
+
+Nodes receiving onion packets may hide their identity from senders by
+"blinding" an arbitrary amount of hops at the end of an onion path.
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of
+the blinded nodes.
+
+They communicate this blinded route and the encrypted blobs to the sender.
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
+may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_data_tlv`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
```

Comment:
As discussed on the BOLT 12 Telegram, it may make sense to get rid of this field since `next_node_id` is currently mandatory, thus making the scid redundant. In addition, IIUC Acinq will currently error if the `next_node_id` field is not present (cc @t-bast).

Reply:
I don't think we should remove it. For payments we still want to use the `short_channel_id` to identify the next node as it uses less space in the onion. For onion messages we're only using `next_node_id`, but for blinded payments we use `short_channel_id`.

Reply:
We can calculate that, but I'd rather just keep this option of using less space, it can always be useful (e.g. to add more dummy hops at the end of a route). This is what we use for normal payments relay, so it shouldn't require implementations to add any new code, so I don't see why we'd want to remove it?

Reply:
> I don't think we should remove it. For payments we still want to use the `short_channel_id` to identify the next node as it uses less space in the onion. For onion messages we're only using `next_node_id`, but for blinded payments we use `short_channel_id`.

That seems fair. Based on that, I'd prefer if we made `next_node_id` explicitly mandatory for onion messages and disallow `scid`, and the reverse requirements for payment relay. (I believe that's what eclair does now, at least the `next_node_id` part?)